### ISSUE TITLE
C# IAsyncStreamReader MoveNext() deadlock

### ISSUE THREAD

**Issue by amitwaisel** [grpc/grpc#8230]
@apolcyn 
When sending a few KBs of data (around 50KB-100KB each) over multiple (10) IServerStreamWriter streams, the task `await stream.MoveNext()` (with/without `CancellationToken`) will never finish (after 10-20 iterations that were completed successfully).
Maybe it is related to the stream buffer? The channel cannot be configured to raise the limit of the stream buffer in C#.

``` C#
List<byte> data = new List<byte>();
// Will run for 10-20 iterations and then freeze
while (await stream.MoveNext())
{
    data.AddRange(stream.Current.Data.ToByteArray());
}
```

**Comment by apolcyn** [comment]
Off the bat I'd wonder if this is causing a memory issue because of the storing in the list, but with those numbers I think memory should be small enough. To check, these messages are uncompressed, correct? (the fix in https://github.com/grpc/grpc/issues/7873 hadn't been released yet).

**Comment by jtattermusch** [comment]
@amitwaisel   we need to know more. What gRPC version? What platform are you on? What you are describing would look like a C core bug, but it seems pretty unlikely because other users do things like this all the time. 

What's the stack trace when it freezes?

**Comment by amitwaisel** [comment]
@apolcyn I didn't check for memory leaks, but as you said - it is small enough for not being noticed.
@jtattermusch We work with the nuget-distributed 1.0.0 version, on Windows platform. The language is C#. Regarding the stack trace, because we use the precompiled library, VisualStudio displays the stack trace up to the MoveNext() call. The task appears to be "scheduled to run".
I found a workaround for this issue, which might indicate what the reason is: Instead of reading stream-by-stream (each one until finished), I created X different tasks that read all streams in parallel. When doing so, everything runs fine.

**Comment by jtattermusch** [comment]
@amitwaisel  my *guess* would be it's one of two possible options:
1.  you are looking at a deadlock in your code. Please note that async continuations for a specific call will be bound to a single specific thread and if some other actions occupy that thread, the continuations cannot run (this should never happen if async/await is used correctly across your code). One thing you can try find the root cause is to set `GrpcEnvironment.SetCompletionQueueCount(1)` (see https://github.com/grpc/grpc/blob/6ea4380d15fa59dd57342a910484ecf7308f31bf/src/csharp/Grpc.Core/GrpcEnvironment.cs#L226). That way, the async continuations will have more threads to run on (but performance may be worse). If that resolves your issue, the problem is most likely in your code.
2. there's an issue in windows version of C core (possible), but to debug this, we would need a minimalist version of your code that reproduces the issue.

**Comment by amitwaisel** [comment]
@jtattermusch I will try to make a simple tester that reproduces the bug. I tried using the `SetCompletionQueueCount(1)` with no noticeable effect. Can it be related to the fact that the client and the server run in the same process [part of testing process]? (For clarification, the issue is related to a stream written (`WriteAsync()`) from the server, and read (`MoveNext()`) from the client).

**Comment by jtattermusch** [comment]
The fact that the client and the server are in the same process tend to lead to a higher risk of deadlock if the user code is written incorrectly (e.g. the use of async/await), but for a correct code, it should absolutely work. Please try to come up with a snippet that reproduces the issue.

**Comment by Ulysess** [comment]
maybe related, use synchronize wait in service implementation will hang, deadlock or something
demonstrate by modified csharp\Grpc.Examples\MathServiceImpl.cs
```
public override async Task Fib(FibArgs request,
	ServerStreamWriter<Num> responseStream,
	ServerCallContext context
) {
	var limit = request.Limit > 0 ? request.Limit : long.MaxValue;
	var fibEnumerator = FibInternal(limit).GetEnumerator();
	while (fibEnumerator.MoveNext()) {
		var task = responseStream.WriteAsync(fibEnumerator.Current);
		task.Wait();  //never return
	}
}
```

**Comment by amitwaisel** [comment]
@Ulysess Using synchronous Wait() inside async function may not be a good idea ([http://blog.stephencleary.com/2012/07/dont-block-on-async-code.html](url))
@jtattermusch Still trying to reproduce this problem out of one of the GRPC examples. It looks that the issue happens when the server stops reading from one of the opened streams (because of an exception not caught by user code and reached the GRPC layer, or just by not reading from the stream). At some point, the client (both C++ and C# in my case) is blocked on **every RPC call** - either writing to any stream using the `WriteAsync` operation (not just the one I don't read from), or even calling another unrelated RPC function.

**Comment by Ulysess** [comment]
Off course i known, only for demonstration.
But it is legal, only not good. Current implementation misbehaved.
Maybe add a new issue for it.

**Comment by ctiller** [comment]
Is this a case of many streams getting cancelled or timing out server side?
It's something we've noticed, and there's a fix incoming (though it's
somewhat complicated).

**Comment by amitwaisel** [comment]
@ctiller  The streams are not getting cancelled or timed out, they are just not being read from (I guess their internal buffers are filled up).
My specific case is that the server writes to X streams (server-to-client), but the client reads only from a single stream (it reads one stream at a time. When it completes one, it goes to the next). In that case, the client completes 20-30 reads (`MoveNext()`) before freezing. Because the logic in one-stream-at-a-time read, it never finishes the first stream, thus not getting to read the rest streams.

**Comment by jtattermusch** [comment]
@amitwaisel   is this still a problem?
